daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala

```diff
@@ -324,6 +326,98 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
       case Nil => throw new IllegalStateException("Not inside a rollback scope")
       case _ :: tailStack => this.copy(rollbackStack = tailStack)
     }
+
+    private def withinRollbackScope: Boolean = rollbackStack.nonEmpty
+
+    /** Let `resolver` be a [[KeyResolver]] that can be used during interpretation to obtain a transaction `tx`
+      * in modes [[com.daml.lf.transaction.ContractKeyUniquenessMode.ContractByKeyUniquenessMode]],
+      * or `Map.empty` in mode [[com.daml.lf.transaction.ContractKeyUniquenessMode.Strict]].
+      * Let `this` state be the result of iterating over `tx` up to a node `n` exclusive using `resolver`.
```

Comment:
Could you please clarify the term `iterate` by referring to the `visitSubtrees` method from the test scope as a reference implementation for the iteration?
(I would have been unable to come up with the iteration approach from `visitSubtrees` just by reading this comment.)

Reply:
Did you read the comment on the `ContractStateMachine` class? That was meant to explain the iteration in detail.
I've nevertheless added references to `visitSubtree` and `Transaction.contractKeyInputs` here and there.